lib/manager/components/CreateFeedSource.js

```diff
@@ -0,0 +1,234 @@
+// @flow
+
+import memoize from 'lodash.memoize'
+import React, {Component} from 'react'
+import update from 'react-addons-update'
+import {
+  Button,
+  Checkbox,
+  Col,
+  ControlLabel,
+  FormControl,
+  FormGroup,
+  Grid,
+  HelpBlock,
+  ListGroup,
+  ListGroupItem,
+  Panel,
+  Row
+} from 'react-bootstrap'
+import {isUri} from 'valid-url'
+
+import ManagerPage from '../../common/components/ManagerPage'
+import ProjectHeader from './ProjectHeader'
+import {validationState} from '../util'
+
+import type {Project, User} from '../../types'
+
+type NewFeed = {
+  autoFetchFeed: boolean,
+  deployable: boolean,
+  name: string,
+  projectId: string,
+  retrievalMethod?: string,
+  url: string
+}
+
+type Props = {
+  createFeedSource: NewFeed => void,
+  fetchProject: (string) => void,
+  project: Project,
+  projectId: string,
+  user: User
+}
+
+type Validation = {
+  _form: boolean,
+  name?: boolean,
+  url?: boolean
+}
+
+type State = {
+  model: NewFeed,
+  validation: Validation
+}
+
+export default class CreateFeedSource extends Component {
+  props: Props
+  state: State
+
+  componentWillMount () {
+    const {fetchProject, project, projectId} = this.props
+    if (!project) {
+      fetchProject(projectId)
+    } else {
+      this._initializeState(this.props)
+    }
+  }
+
+  componentWillReceiveProps (nextProps: Props) {
+    this._initializeState(nextProps)
+  }
+
+  _initializeState (props: Props) {
+    this.setState({
+      model: {
+        autoFetchFeed: false,
+        deployable: false,
+        name: '',
+        projectId: props.projectId,
+        url: ''
+      },
+      validation: {
+        _form: false,
+        name: true,
+        url: true
+      }
+    })
+  }
+
+  _onInputChange = memoize(
+    (fieldName: string) => (evt: Event & {target: HTMLInputElement}) => {
+      const updatedState: State = update(this.state, {
+        model: {[fieldName]: {$set: evt.target.value}}
+      })
+      this.setState(updatedState)
+      this._validateModel(updatedState.model)
+    }
+  )
+
+  _onSave = () => {
+    const {model} = this.state
+    if (model.url === '') {
+      delete model.url
+    }
+    model.retrievalMethod = model.autoFetchFeed
+      ? 'FETCHED_AUTOMATICALLY'
+      : 'MANUALLY_UPLOADED'
+    this.props.createFeedSource(model)
+  }
+
+  _toggleCheckBox = memoize((fieldName: string) => () => {
+    this.setState(
+      update(this.state, {
+        model: {[fieldName]: {$set: !this.state.model[fieldName]}}
+      })
+    )
+  })
+
+  _validateModel (model: NewFeed) {
+    const validation: Validation = {
+      _form: false,
+      name: !(!model.name || model.name.length === 0),
+      url: model.url === '' || !!isUri(model.url)
+    }
+    validation._form = !!(validation.name && validation.url)
+    this.setState({ validation })
+  }
+
+  render () {
```

Comment:
Why not do something similar to CreateProject's use of GeneralSettings and include a shared component here?  It seems like that approach would be more maintainable (and reduce duplicated code).

Reply:
It is different enough that it would get too unwieldy to account for all the differences.  The main difference is that all of the fields in the existing FeedSourceSettings component automatically save upon change.  In the create version, nothing is saved until you hit the save button.